README.md
# Backend of the xRAM-Memory website

This project concentrates the content management system and the API for the xRAM-Memory website. The project used several Python libraries, especially: Django (base framework and CMS), Django REST Framework (API), newspaper3k and goose3 (news crawlers), django-filer (file management) and celery (queue management for processing). distributed).

Project dependencies are managed with the `pipenv` tool.

## Instalação

0. Install [Python 3.7.\*](https://www.python.org/downloads/) and [Pipenv](https://pypi.org/project/pipenv/)
1. Install the packages: `python3-dev(el)`, `python3-pdfkit`, `poppler-utils` and `wkhtmltopdf`
2. Install project dependencies: `pipenv install --dev --python 3.7` and `npm install`
3. Set environment variables in `.env` from `.env.example`, in particular, set at least `DJANGO_SECRET_KEY`, `DJANGO_LUNR_INDEX_REMOTE_SECRET` and `DJANGO_HASHID_FIELD_SALT` variables
4. Enter pipenv shell: `pipenv shell`
5. Install nltk bodies: `python ./scripts/download_corpora.py --user`
6. Collect the statics: `./manage.py collectstatic`
7. Create the media dir and the lunr index: `mkdir -p ./media/lunr_index && ./media/lunr_index/index.json`
8. Run the migrations: `./manage.py migrate`
9. Create a superuser: `./manage.py createsuperuser`

## Local environment

### Application

0. Run the installation steps above
1. Enter a `pipenv` shell
2. Start a memcached container: `docker run --name my-memcache -d -p 127.0.0.1:11211:11211 memcached memcached -m 64`
3. Start the application locally: `./manage.py runserver_plus`
4. Go to `http://localhost:8000/admin/`
5. Clone, build the image and run the [`micro-lunr_index_builder` project](https://github.com/felubra/micro-lunr_index_builder):

```shell
docker run -d --name xram_memory__lunr_index_builder -v <absolute path to the index file created on the step 7 above>:/usr/src/app/index.json:rw -e AUTH_TOKEN=<DJANGO_LUNR_INDEX_REMOTE_SECRET value> -p 3001:3000 xram_memory/lunr_index_builder
```

### Celery

0. Start a redis container listening on the default port locally:

```shell
docker run --name some-redis -d -p 127.0.0.1:6379:6379 redis
```

1. Start as many instances of celery as necessary. In a `pipenv` shell, run:

```shell
celery worker -A xram_memory -n 1
```

3. (Optional) Monitor workers with _Flower_. In a `pipenv` shell, run:

```shell
flower -A xram_memory
```

## Data Entities

The following models (entities) are present in this project:

The diagram below illustrates the relationship between the entities (click to open the image):

[![diagram](./docs/entidades.png)](./docs/entidades.png)

## Project structure

A Django project is made up of several applications, each with a specific responsibility, which can be expressed by the organization of folders. This is the structure of the `./xram_memory` folder, which contains the system source code:

```
.
├── albums - Album management application
├── artifact - Artifact management application
├── lib - Specialized libraries developed for the project
│ ├── file_previews - File preview generator
│ ├── news_fetcher - Main system program, more about it below
│ │ └── plugins - The various plugin implementations
│ │ ├── archives - News Archivers
│ │ ├── parsers - News Content Extractors
│ │ └── pdf_captures - PDF news capture generators
│ └── stopwords - A dictionary of stopwords in multiple languages
├── logger - Application responsible for logging system operations
├── lunr_index - Application responsible for generating client-side search indexes
├── page - Application for managing Static Pages
├── quill_widget - Application that provides a rich text editing widget with the Quill.js library
├── search_indexes - Application responsible for generating server-side search indexes (ElasticSearch)
├── static - Static files
├── taxonomy - Application responsible for classifying content (Taxonomy)
├── templates - Global changes to templates
├── users - Application responsible for managing users and groups
└── utils - Global Utilities
```

In the `./tests` folder we have the project tests. Code coverage is currently at 65%.


docs/CHANGELOG.md
# Changelog

## 1.5.0

Adicionado

```
- News listing endpoint (#160) [Felipe Lube de Bragança]

  * feat: install django rest framewor's token authentication

  * feat: new endpoint to get all the news info for static generation

  * feat: auth token endpoint

  * fix: update readme

Corrigido
~~~~~~~~~
- Rename file. [Felipe]

Outros
~~~~~~
- Chore(deps): bump pillow from 8.2.0 to 8.3.2 (#158) [dependabot[bot]]

  Bumps [pillow](https://github.com/python-pillow/Pillow) from 8.2.0 to 8.3.2.
  - [Release notes](https://github.com/python-pillow/Pillow/releases)
  - [Changelog](https://github.com/python-pillow/Pillow/blob/master/CHANGES.rst)
  - [Commits](https://github.com/python-pillow/Pillow/compare/8.2.0...8.3.2)

  ---
  updated-dependencies:
  - dependency-name: pillow
    dependency-type: direct:production
  ...
- Chore(deps): bump sqlparse from 0.4.1 to 0.4.2 (#159)
  [dependabot[bot]]

  Bumps [sqlparse](https://github.com/andialbrecht/sqlparse) from 0.4.1 to 0.4.2.
  - [Release notes](https://github.com/andialbrecht/sqlparse/releases)
  - [Changelog](https://github.com/andialbrecht/sqlparse/blob/master/CHANGELOG)
  - [Commits](https://github.com/andialbrecht/sqlparse/compare/0.4.1...0.4.2)

  ---
  updated-dependencies:
  - dependency-name: sqlparse
    dependency-type: indirect
  ...
- Chore(deps): bump pillow from 8.1.2 to 8.2.0 (#155) [dependabot[bot]]

  Bumps [pillow](https://github.com/python-pillow/Pillow) from 8.1.2 to 8.2.0.
  - [Release notes](https://github.com/python-pillow/Pillow/releases)
  - [Changelog](https://github.com/python-pillow/Pillow/blob/master/CHANGES.rst)
  - [Commits](https://github.com/python-pillow/Pillow/compare/8.1.2...8.2.0)

  ---
  updated-dependencies:
  - dependency-name: pillow
    dependency-type: direct:production
  ...
- Chore(deps): bump django from 2.2.20 to 2.2.24 (#157)
  [dependabot[bot]]

  Bumps [django](https://github.com/django/django) from 2.2.20 to 2.2.24.
  - [Release notes](https://github.com/django/django/releases)
  - [Commits](https://github.com/django/django/compare/2.2.20...2.2.24)

  ---
  updated-dependencies:
  - dependency-name: django
    dependency-type: direct:production
  ...
- Chore(deps): bump django from 2.2.19 to 2.2.20 (#151)
  [dependabot[bot]]

  Bumps [django](https://github.com/django/django) from 2.2.19 to 2.2.20.
  - [Release notes](https://github.com/django/django/releases)
  - [Commits](https://github.com/django/django/compare/2.2.19...2.2.20)
- Chore(deps): [security] bump django-debug-toolbar from 1.10.1 to
  1.11.1 (#152) [dependabot-preview[bot]]

  Bumps [django-debug-toolbar](https://github.com/jazzband/django-debug-toolbar) from 1.10.1 to 1.11.1. **This update includes a security fix.**
  - [Release notes](https://github.com/jazzband/django-debug-toolbar/releases)
  - [Changelog](https://github.com/jazzband/django-debug-toolbar/blob/1.11.1/docs/changes.rst)
  - [Commits](https://github.com/jazzband/django-debug-toolbar/compare/1.10.1...1.11.1)


1.4.1 (2021-03-28)
------------------

Adicionado
```

- Abort app start if there is access error in lunr index folder. [Felipe
  Lube de Bragança]
- Allow download corpora to user folder. [Felipe Lube de Bragança]
- Otimizações no endpoint dos assuntos - Utilize um serializer mais
  simples e rápido para a maior dos casos - Obtenha objetos relacionados
  e guarde-os em cache - Somente pegue assuntos que tenham algum item
  relacionado - Remova campo não utilizado em SubjectSerializer. [Felipe
  Lube de Bragança]

Corrigido

```
- Revert django-elasticsearch-dsl update. [Felipe Lube de Bragança]
- Fix (Pipfile): use compatible release indicator. [Felipe Lube de
  Bragança]
- Fix(docs): add missing command to README.md. [Felipe Lube de Bragança]
- Get_file_icon: raise exception if no icon file is found. [Felipe Lube
  de Bragança]

Outros
~~~~~~
- Chore: update Pillow. [Felipe Lube de Bragança]
- Chore: update packages to avoid security issues. [Felipe Lube de
  Bragança]
- Chore: update cairosvg. [Felipe Lube de Bragança]
- Chore: update to DJango 2.2.19. [Felipe Lube de Bragança]
- Chore(docs): update readme with run instructions. [Felipe Lube de
  Bragança]
- Refactor: use try blocks in signal stopping decorators. [Felipe Lube
  de Bragança]
- Chore: readme: update install instructions. [Felipe Lube de Bragança]
- Chore(docs): update readme. [Felipe Lube de Bragança]
- Refactor: download_corpora: use boolean argument --user. [Felipe Lube
  de Bragança]
- Chore(deps): [security] bump djangorestframework from 3.11.1 to 3.11.2
  (#149) [dependabot-preview[bot]]

  Bumps [djangorestframework](https://github.com/encode/django-rest-framework) from 3.11.1 to 3.11.2. **This update includes a security fix.**
  - [Release notes](https://github.com/encode/django-rest-framework/releases)
  - [Commits](https://github.com/encode/django-rest-framework/compare/3.11.1...3.11.2)
- Chore(deps): [security] bump lxml from 4.5.2 to 4.6.2 (#146)
  [dependabot-preview[bot]]

  Bumps [lxml](https://github.com/lxml/lxml) from 4.5.2 to 4.6.2. **This update includes a security fix.**
  - [Release notes](https://github.com/lxml/lxml/releases)
  - [Changelog](https://github.com/lxml/lxml/blob/master/CHANGES.txt)
  - [Commits](https://github.com/lxml/lxml/compare/lxml-4.5.2...lxml-4.6.2)
- Refactor: um gerenciador de contexto não precisa ser uma fixture.
  [Felipe Lube de Bragança]
- Chore: limpeza nas importações e remoção de arquivos inúteis. [Felipe
  Lube de Bragança]
- Refactor: Não filtre a saída do endpoint das palavras-chave. [Felipe
  Lube de Bragança]

  BREAKING CHANGE

  A responsabilidade pela filtragem das palavras-chave irrelevantes é de quem as insere no banco de dados
- Merge branch 'dev' of github.com:felubra/django-xram_memory into dev.
  [Felipe Lube de Bragança]
- Reestilização do formulário p/ adicionar várias notícias (#142)
  [Felipe Lube de Bragança]

  * chore: atualização no arquivo .env de exemplo

  * feat: breadcrumbs


xram_memory/settings.py
"""
Django settings for xram_memory project.

For more information on this file, see
https://docs.djangoproject.com/en/2.1/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/2.1/ref/settings/
"""
import os

from configurations import Configuration, values
from xram_memory.lunr_index.util import LunrBackendValue


class Common(Configuration):
    # Build paths inside the project like this: os.path.join(BASE_DIR, ...)
    BASE_DIR = os.path.dirname(os.path.dirname(__file__))

    # SECURITY WARNING: keep the secret key used in production secret!
    SECRET_KEY = values.SecretValue()

    # SECURITY WARNING: don't run with debug turned on in production!
    DEBUG = values.BooleanValue(False)

    ALLOWED_HOSTS = values.ListValue([])

    # Application definition
    INSTALLED_APPS = [
        "xram_memory.apps.DefaultAdminConfig",
        "django.contrib.auth",
        "django.contrib.contenttypes",
        "django.contrib.sessions",
        "django.contrib.messages",
        "whitenoise.runserver_nostatic",
        "django.contrib.staticfiles",
        "django_extensions",
        "xram_memory.users",
        "xram_memory.taxonomy",
        "xram_memory.logger",
        "xram_memory.quill_widget",
        "xram_memory.artifact",
        "xram_memory.page",
        "easy_thumbnails",
        "xram_memory.apps.FilerConfig",
        "mptt",
        "rest_framework",
        "rest_framework.authtoken",
        "corsheaders",
        "tags_input",
        "xram_memory.albums",
    ]

    MIDDLEWARE = [
        "corsheaders.middleware.CorsMiddleware",
        "django.middleware.security.SecurityMiddleware",
        "whitenoise.middleware.WhiteNoiseMiddleware",
        "django.contrib.sessions.middleware.SessionMiddleware",
        "django.middleware.common.CommonMiddleware",
        "django.middleware.csrf.CsrfViewMiddleware",
        "django.contrib.auth.middleware.AuthenticationMiddleware",
        "django.contrib.messages.middleware.MessageMiddleware",
        "django.middleware.clickjacking.XFrameOptionsMiddleware",
        "django_currentuser.middleware.ThreadLocalUserMiddleware",
    ]

    ROOT_URLCONF = "xram_memory.urls"

    TEMPLATES = [
        {
            "BACKEND": "django.template.backends.django.DjangoTemplates",
            "DIRS": [os.path.join(BASE_DIR, "xram_memory", "templates")],
            "APP_DIRS": True,
            "OPTIONS": {
                "context_processors": [
                    "django.template.context_processors.debug",
                    "django.template.context_processors.request",
                    "django.contrib.auth.context_processors.auth",
                    "django.contrib.messages.context_processors.messages",
                ],
            },
        },
    ]

    STATICFILES_DIRS = [
        os.path.join(BASE_DIR, "xram_memory", "static"),
    ]

    WSGI_APPLICATION = "xram_memory.wsgi.application"

    # Database
    # https://docs.djangoproject.com/en/2.1/ref/settings/#databases
    DATABASES = values.DatabaseURLValue(
        "sqlite:///{}".format(os.path.join(BASE_DIR, "db.sqlite3"))
    )

    # Password validation
    # https://docs.djangoproject.com/en/2.1/ref/settings/#auth-password-validators
    AUTH_PASSWORD_VALIDATORS = [
        {
            "NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator",
        },
        {
            "NAME": "django.contrib.auth.password_validation.MinimumLengthValidator",
        },
        {
            "NAME": "django.contrib.auth.password_validation.CommonPasswordValidator",
        },
        {
            "NAME": "django.contrib.auth.password_validation.NumericPasswordValidator",
        },
    ]

    # Internationalization
    # https://docs.djangoproject.com/en/2.1/topics/i18n/
    LANGUAGE_CODE = "pt-br"

    TIME_ZONE = "America/Sao_Paulo"

    USE_I18N = True

    USE_L10N = True

    USE_TZ = True

    # Static files (CSS, JavaScript, Images)
    # https://docs.djangoproject.com/en/2.1/howto/static-files/
    STATIC_URL = "/static/"
    STATIC_ROOT = os.path.join(BASE_DIR, "staticfiles")
    STATICFILES_STORAGE = (
        "xram_memory.utils.PatchedCompressedManifestStaticFilesStorage"
    )

    AUTH_USER_MODEL = "users.User"

    RQ_QUEUES = {
        "default": {
            "HOST": "localhost",
            "PORT": 6379,
            "DB": 0,
            "DEFAULT_TIMEOUT": 360,
            "ASYNC": False,
        },
    }
    MEDIA_ROOT = os.path.join(BASE_DIR, "media")
    MEDIA_URL = "/media/"
    VALID_FILE_UPLOAD_MIME_TYPES = (
        "image/jpeg",
        "image/png",
        "image/gif",
        "image/webp",
        "application/pdf",
        "application/zip",
    )
    VALID_FILE_UPLOAD_IMAGES_MIME_TYPES = (
        "image/jpeg",
        "image/png",
        "image/gif",
        "image/webp",
    )

    THUMBNAIL_SOURCE_GENERATORS = (
        "easy_thumbnails.source_generators.pil_image",
        "xram_memory.lib.file_previews.pdf_preview",
        "xram_memory.lib.file_previews.icon_preview",
    )

    THUMBNAIL_ALIASES = {
        "": {
            "document_preview": {
                "size": (850, 850),
                "autocrop": False,
                "crop": "scale",
                "upscale": False,
            },
            "document_thumbnail": {
                "size": (0, 250),
                "autocrop": False,
                "crop": "scale",
                "upscale": False,
            },
            "news_page": {
                "size": (0, 350),
                "autocrop": True,
                "crop": "smart",
                "upscale": False,
            },
            "thumbnail": {
                "size": (250, 250),
                "background": "#f3f1f1",
                "autocrop": True,
                "crop": "smart",
            },
            "image_capture": {"size": (670, 204), "autocrop": True, "crop": "scale"},
            # TODO: desabilitar?
            "1280": {"size": (1280, 1280), "crop": "scale"},
            "640": {"size": (640, 360), "crop": "scale"},
            "360": {"size": (360, 360), "crop": "scale"},
            "favicon": {
                "size": (18, 18),
                "autocrop": True,
                "scale_and_crop": "smart",
                "upscale": False,
            },
        }
    }

    STATICFILES_FINDERS = [
        "django.contrib.staticfiles.finders.FileSystemFinder",
        "django.contrib.staticfiles.finders.AppDirectoriesFinder",
        "npm.finders.NpmFinder",
    ]
    NPM_ROOT_PATH = BASE_DIR
    NPM_FILE_PATTERNS = {
        "file-icon-vectors": [os.path.join("dist", "icons", "vivid", "*")],
        "stopwords-iso": ["stopwords-iso.json"],
        "material-design-icons": [
            os.path.join("navigation", "svg", "production", "ic_fullscreen*"),
            os.path.join("action", "svg", "production", "ic_info_24px*"),
            os.path.join("image", "svg", "production", "ic_picture_as_pdf_24px*"),
            os.path.join("image", "svg", "production", "ic_filter_24px*"),
        ],
        "quill": [os.path.join("dist", "*")],
        "screenfull": [os.path.join("dist", "*")],
    }

    REST_FRAMEWORK = {
        "DEFAULT_PARSER_CLASSES": ("rest_framework.parsers.JSONParser",),
        "DEFAULT_THROTTLE_CLASSES": ("rest_framework.throttling.AnonRateThrottle",),
        "DEFAULT_AUTHENTICATION_CLASSES": [
            "rest_framework.authentication.TokenAuthentication",
        ],
        "DEFAULT_THROTTLE_RATES": {
            "anon": "60/minute",
        },
    }

    FILER_FILE_MODELS = ["artifact.Document"]

    TAGS_INPUT_MAPPINGS = {
        "taxonomy.Keyword": {
            "field": "name",
            "create_missing": True,
        },
        "taxonomy.Subject": {
            "field": "name",
            "create_missing": True,
        },
    }
    FILE_UPLOAD_MAX_MEMORY_SIZE = 10621440
    CELERY_BROKER_URL = values.Value("")
    CELERY_CELERYD_MAX_TASKS_PER_CHILD = 10

    FOLDER_CAPTURES = {"name": "Capturas automáticas", "id": 1}

    FOLDER_PHOTO_ALBUMS = {"name": "Álbuns de fotos", "id": 2}

    FOLDER_PDF_CAPTURES = {"name": "Capturas de notícias em PDF", "id": 3}

    FOLDER_IMAGE_CAPTURES = {"name": "Imagens de notícias", "id": 4}

    DEFAULT_FOLDERS = (
        FOLDER_CAPTURES,
        FOLDER_PHOTO_ALBUMS,
        FOLDER_PDF_CAPTURES,
        FOLDER_IMAGE_CAPTURES,
    )

    HASHID_FIELD_SALT = values.Value("4OkZKanuMWUU4EO92FcDmwXkn6PbksGAIClUcG4S")
    HASHID_FIELD_LOOKUP_EXCEPTION = False
    HASHID_FIELD_ALLOW_INT_LOOKUP = False
    # Sal usado para gerar um nome de arquivo nas funções de captura em News
    FILE_HASHING_SALT = values.Value("hs204ViIUpIu45CTTUl3KsoQJgVtnmrHpXvRl8u5")

    CACHES = {
        "default": {
            "BACKEND": "django.core.cache.backends.memcached.MemcachedCache",
            "LOCATION": values.Value(
                "127.0.0.1:11211", True, environ_name="MEMCACHED_URL"
            ),
        }
    }


class IndexingWithElasticSearch(Common):
    # Application definition
    INSTALLED_APPS = Common.INSTALLED_APPS + [
        "django_elasticsearch_dsl",
        "xram_memory.search_indexes",
    ]

    ELASTICSEARCH_INDEX_NAMES = {
        "xram_memory.search_indexes.documents.news": "artifact_news",
        "xram_memory.search_indexes.documents.document": "artifact_document",
    }

    ELASTICSEARCH_DSL = {
        "default": {
            "hosts": values.Value(
                "localhost:9200",
                True,
                environ_name="ELASTICSEARCH_HOST",
                environ_prefix="DJANGO",
            ),
            "http_auth": values.TupleValue(
                environ_name="ELASTICSEARCH_CREDENTIALS", environ_prefix="DJANGO"
            ),
            "timeout": 30,
        },
    }

    ELASTICSEARCH_DSL_SIGNAL_PROCESSOR = (
        "xram_memory.search_indexes.signals.FixtureAwareSignalProcessor"
    )


class IndexingWithLunrSearch(Common):
    # Application definition
    INSTALLED_APPS = Common.INSTALLED_APPS + [
        "xram_memory.lunr_index",
    ]

    # Intervalo mínimo entre a criação de arquivos de índice do Lunr
    LUNR_INDEX_REBUILD_INTERVAL = values.IntegerValue(10 * 60)  # 10 minutos
    # Tempo máximo que a operação deve levar, após o qual falhará
    LUNR_INDEX_REBUILD_TIMEOUT = values.IntegerValue(5 * 60)  # 5 minutos
    # Se devemos salvar o documento no indice - somente suportado pelo Elastic Lunr
    LUNR_INDEX_SAVE_DOCUMENT = values.BooleanValue(True)
    # Quais campos dos modelos indexar para busca
    LUNR_INDEX_SEARCH_FIELDS = values.ListValue(["title", "teaser"])
    # Caminho do arquivo do índice relativo a MEDIA_ROOT
    LUNR_INDEX_FILE_PATH = values.Value(
        os.path.join(Common.MEDIA_ROOT, "lunr_index/index.json")
    )
    # Tipo de reconstrução: 'local' (usa lunr.py para gerar localmente) ou 'remote' envia requisição http com dados
    # a serem indexados para o servidor definido em REMOTE_HOST

    # FIXME: exigência condicional de variável não funciona, porque o valor não pode ser lido instantaneamente aqui;
    # alternativa: usar uma connection string: file:///path/to/the/file para processamento local ou
    # http://secret@host:port para processamento remoto, eliminando as variáveis LUNR_INDEX_BACKEND,
    # LUNR_INDEX_FILE_PATH, LUNR_INDEX_REMOTE_HOST e LUNR_INDEX_REMOTE_SECRET
    LUNR_INDEX_BACKEND = LunrBackendValue(environ_required=True)
    # Host para onde as instâncias dos modelos a serem indexados devem ser enviados
    LUNR_INDEX_REMOTE_HOST = values.Value(
        environ_required=LUNR_INDEX_BACKEND == LunrBackendValue.BACKEND_REMOTE
    )
    # Segredo usado para autenticação http Bearer Token no servidor REMOTE_HOST
    LUNR_INDEX_REMOTE_SECRET = values.Value(
        environ_required=LUNR_INDEX_BACKEND == LunrBackendValue.BACKEND_REMOTE
    )


class IndexingWithAllApps(IndexingWithElasticSearch, IndexingWithLunrSearch, Common):
    # Application definition
    INSTALLED_APPS = Common.INSTALLED_APPS + [
        "django_elasticsearch_dsl",
        "xram_memory.search_indexes",
        "xram_memory.lunr_index",
    ]


class Development(IndexingWithAllApps):
    """
    The in-development settings and the default configuration.
    """

    DEBUG = True

    ALLOWED_HOSTS = values.ListValue(
        [
            "localhost",
            "127.0.0.1",
            "[::1]",
            "192.168.99.100",
            "xram-memory.local",
            "xram-memory.localhost",
        ]
    )

    INTERNAL_IPS = ["127.0.0.1"]

    MIDDLEWARE = Common.MIDDLEWARE + ["debug_toolbar.middleware.DebugToolbarMiddleware"]

    INSTALLED_APPS = IndexingWithAllApps.INSTALLED_APPS + [
        "debug_toolbar",
        "cache_fallback",
    ]

    LOGGING = {
        "version": 1,
        "disable_existing_loggers": False,
        "handlers": {
            "console": {
                "class": "logging.StreamHandler",
            },
        },
        "loggers": {
            "": {
                "handlers": ["console"],
                "level": "ERROR",
            },
            "django": {
                "handlers": ["console"],
                "level": "INFO",
            },
        },
    }
    CELERY_BROKER_URL = values.Value("redis://127.0.0.1:6379/0")

    CORS_ORIGIN_ALLOW_ALL = True


class DevelopmentWithDocker(Development):
    # Habilite o debug toolbar no ambiente docker
    # https://gist.github.com/douglasmiranda/9de51aaba14543851ca3
    DEBUG_TOOLBAR_CONFIG = {
        "SHOW_TOOLBAR_CALLBACK": "xram_memory.utils.show_toolbar",
    }


class Staging(IndexingWithLunrSearch):
    """
    The in-staging settings.
    """

    # Security
    SESSION_COOKIE_SECURE = values.BooleanValue(True)
    SECURE_BROWSER_XSS_FILTER = values.BooleanValue(True)
    SECURE_CONTENT_TYPE_NOSNIFF = values.BooleanValue(True)
    SECURE_HSTS_INCLUDE_SUBDOMAINS = values.BooleanValue(True)
    SECURE_HSTS_SECONDS = values.IntegerValue(3600)
    SECURE_REDIRECT_EXEMPT = values.ListValue([])
    SECURE_SSL_HOST = values.Value(None)
    SECURE_SSL_REDIRECT = values.BooleanValue(True)
    SECURE_PROXY_SSL_HEADER = values.TupleValue(("HTTP_X_FORWARDED_PROTO", "https"))

    ALLOWED_HOSTS = values.ListValue(
        default=["xram-memory.felipelube.com", "xram-memory.localhost"]
    )
    USE_X_FORWARDED_HOST = True
    SECURE_PROXY_SSL_HEADER = ("HTTP_X_FORWARDED_PROTO", "https")

    LOGGING = {
        "version": 1,
        "disable_existing_loggers": False,
        "handlers": {
            "console": {
                "class": "logging.StreamHandler",
            },
        },
        "loggers": {
            "": {
                "handlers": ["console"],
                "level": "ERROR",
            },
        },
    }

    CACHES = {
        "default": {
            "BACKEND": "django.core.cache.backends.memcached.MemcachedCache",
            "LOCATION": values.Value(
                "127.0.0.1:11211", True, environ_name="MEMCACHED_URL"
            ),
        }
    }


class Production(Staging):
    """
    The in-production settings.
    """
